File: js/src/gilded_rose.js
type Item {
  name: string;
  sell_in: number;
  quality: number;
}

function Item(name, sell_in, quality) {
  this.name = name;
  this.sell_in = sell_in;
  this.quality = quality;
}

function extractByName(name:string, items:Array<Item>):Array<Item> {
  // ...
}

function withoutByName(name:string, items:Array<Item>):Array<Item> {
  // ...
}

function updateBackstagePasses(items:Array<Item>):Array<Item> {
  // ...
}

function updateLegendaryItems(items:Array<Item>):Array<Item> {
  // ...
}

function updateAgedBrieItems(items:Array<Item>):Array<Item> {
  // ...
}

function updateConjuredItems(items:Array<Item>):Array<Item> {
  // ...
}

function updateGeneralItems(items:Array<Item>):Array<Item> {
  // ...
}

function updateQuality(items:Array<Item>):Array<Item> {
  var backstagePasses = extractByName('Backstage passes to a TAFKAL80ETC concert', items)
  backstagePasses = updateBackstagePasses(backstagePasses)
  items = withoutByName('Backstage passes to a TAFKAL80ETC concert', items)

  var brie = extractByName('Aged Brie', items)
  brie = updateAgedBrieItems(brie)
  items = withoutByName('Aged Brie')

  // ... do the other special ones, blah blah

  items = updateGeneralItems(items)

  return concat([], backstagePasses, brie, items)
}

File: ruby/gilded_rose.rb
class GildedRose

  def initialize items
    @items = items
  end

  def update_quality
    @items.each do |item|
      case item.name
      when "Aged Brie"
        AgedBrie.new(item).update
      when "Sulfuras, Hand of Ragnaros"
        Sulfuras.new(item).update
      when "Backstage passes to a TAFKAL80ETC concert"
        Backstage.new(item).update
      else
        Default.new(item).update
      end
    end
  end

  class Base
    def initialize item
      @item = item
    end
  end

  class Backstage < Base
    def update
      @item.sell_in -= 1
      @item.quality += 1

      @item.quality += 1 if @item.sell_in <= 10
      @item.quality += 1 if @item.sell_in <= 5
      @item.quality = 0 if @item.sell_in <= 0
    end
  end

  class Sulfuras < Base
    def update
      @item.sell_in -= 1
    end
  end

  class AgedBrie < Base
    def update
      @item.sell_in -= 1
      @item.quality += 1 unless @item.quality == 50

      if @item.sell_in <= 0
        @item.quality += 1 unless @item.quality == 50
      end
    end
  end

  class Default < Base
    def update
      @item.sell_in -= 1
      @item.quality -= 1 unless @item.quality == 0

      if @item.sell_in <= 0
        @item.quality -= 1 unless @item.quality == 0
      end
    end
  end

  GILDED_CLASSES = {
    "Aged Brie" => AgedBrie,
    "Sulfuras, Hand of Ragnaros" => Sulfuras,
    "Backstage passes to a TAFKAL80ETC concert" => Backstage
  }

end

class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end
